qiskit_experiments/curve_analysis/models.py

```diff
@@ -0,0 +1,354 @@
+
+import time
+import warnings
+from typing import Callable, List, Dict, Tuple, Sequence, Iterator
+
+import numpy as np
+import uncertainties
+from asteval import Interpreter, get_ast_names
+from lmfit.minimizer import Minimizer
+from lmfit.model import Model
+from lmfit.parameter import Parameters
+
+from qiskit_experiments.curve_analysis.curve_data import SeriesDef, SolverResult
+from qiskit_experiments.exceptions import AnalysisError
+
+
+class CurveModel(Model):
+    """A model that describes the curve fitting function.
+
+    The model manages generation of the fitting function, computation of
+    residual value against given parameters as a cost function of optimization,
+    and preparation of parameter objects passed to the solver instance.
+    The :class:`CurveModel` is implemented on top of the `LMFIT`_ ``Model`` class
+    with some extension for the multi objective optimization
+    described in :ref:`curve_analysis_overview`.
+
+    The CurveModel is instantiated with a list of :class:`.SeriesDef` instances,
+    that may provide a string representation of the fitting function
+    along with some metadata to manage data processing and visualization.
+
+    The fitting function description is parsed with the `ASTEVAL`_ parser, which
+    evaluates mathematical expressions. Note that mathematical functions are
+    imported only from Python's math module and numpy for the safety on code execution.
+    This mechanism allows us to easily serialize an analysis subclass instance
+    that is initialized with fitting functions, i.e. serialization of callables is not necessary.
+    In addition, the function description is now human-readable string,
+    and thus it can be directly copied to analysis result for helping experimentalists to
+    investigate the fitting outcome.
+
+    Fitting parameters are also automatically extracted from the function description,
+    and thus an analysis class author just needs to provide a function description
+    when one wants to define new analysis class.
+
+    Here is an example of acceptable function descriptions.
+
+    .. parsed-literal::
+
+        amp * sin(2 * pi * freq * x + phi) + base
+        amp * cos(2 * pi * freq * x + phi) + base
+
+    The parameter name ``x`` is reserved by the model to represent a parameter to be scanned.
+    Thus, a composite model comprising two functions above consists of four independent parameters
+    ``amp``, ``freq``, ``phi``, and ``base`` that will be evaluated by the curve fitter.
+    Other names ``cos``, ``sin``, and ``pi`` are universal functions and
+    a constant value imported from the Python's math module in this example.
+    Note that when multiple series definitions are provided like this example,
+    the parameters with the same name are shared among the fit functions.
+
+    To compute the residual value, input and output values of each function are concatenated.
```

Comment:
What does concatenated mean in this context?

Reply:
Minimizer requires 1d array rather than N-d array for N cost functions. Anyways this is removed.